src/core/profiling/endoscope_backend.h

```diff
@@ -0,0 +1,163 @@
+#define marker_capacity 10000
+#define task_capacity 10000
+#define atom_capacity 50000
+#define thread_capacity 100
+
+/* Types */
+
+typedef struct /* EndoMarker */ {  /* 47/48 marker_id as array index */
+  ENDO_INT64 cycle_created;  /* overflow after 83 years on 3.5GHz CPU */
+  ENDO_INT64 timestamp;
+  const char *name;
+  const char *file;
+  const char *function_name;
+  int line;
+  ENDO_INDEX next_marker;  /* for hash map collision */
+  unsigned char type;  /* EndoMarkerPB::MarkerType */
+} EndoMarker;
+
+typedef struct /* EndoTask */ {  /* 32/32 */
+  ENDO_INT64 cycle_begin;
+  ENDO_INT64 cycle_end;
+  unsigned int task_id;
+  ENDO_INDEX marker_id;
+  ENDO_INDEX thread_index;
+  ENDO_INDEX log_head;
+  ENDO_INDEX log_tail;
+  ENDO_INDEX next_task;
+  ENDO_INDEX next_taskwithatom;
+  unsigned short scope_depth;  /* max depth 65535 */
+} EndoTask;
+
+typedef struct /* EndoAtom */ {  /* 13/16 */
+  ENDO_INT64 cycle;
+  ENDO_INDEX param;  /* marker_id if ECOPE_BEGIN/EVENT/ERROR */
+  ENDO_INDEX next_atom;  /* for both task log and available atoms */
+  unsigned char type;  /* EndoAtomPB::AtomType */
+} EndoAtom;
+
+typedef struct /* EndoThread */ {  /* 22/24 */
+  ENDO_INT64 cycle_created;
+  ENDO_INT64 timestamp;
+  int thread_id;
+  ENDO_INDEX task_active;
+} EndoThread;
+
+typedef struct /* EndoBase */ {  /* endoscope instance */
+  EndoMarker marker_pool[marker_capacity + 1];
+  ENDO_INDEX marker_count;  /* number of used markers */
+  ENDO_INDEX marker_map[hash_size];  /* lookup table for used markers */
+
+  EndoTask task_pool[task_capacity + 1];
+  ENDO_INDEX task_stack;  /* next available task */
+  ENDO_INDEX task_history_head;
+  ENDO_INDEX task_history_tail;
+  ENDO_INDEX task_withatom_head;
+  ENDO_INDEX task_withatom_tail;
+  unsigned int task_count;  /* for task_id only, old task may be deleted */
+
+  EndoAtom atom_pool[atom_capacity];  /* available-stack */
+  ENDO_INDEX atom_stack;  /* next available atom */
+
+  EndoThread thread_pool[thread_capacity + 1];
+  ENDO_INDEX thread_count;
+
+  EndoMarker *marker_error;
+  EndoTask *task_error;
+  EndoThread *thread_error;
+
+  char errormsg[25];
+
+  ENDO_INT64 cycle_begin;
+  ENDO_INT64 cycle_sync;
+  double time_begin;
+  double time_sync;
+
+  void* mutex;  /* pointer to mutex for this instance */
+} EndoBase;
+
+/* function definition */
+
+void endoscope_init(EndoBase *base);
+
+void endoscope_destroy(EndoBase *base);
+
+ENDO_INT64* endoscope_begin(EndoBase *base, const char *name,
+    const char *file, const int line, const char* function_name);
+
+void endoscope_end(EndoBase *base, const char *name, ENDO_INT64 cycle_end);
+
+void endoscope_event(EndoBase *base, const char *name,
+    const char *file, const int line, const char* function_name,
+    ENDO_INT64 cycle_event);
+
+void endoscope_error(EndoBase *base, const char *name,
+    const char *file, const int line, const char* function_name,
+    ENDO_INT64 cycle_event);
+
+ENDO_INT64 endoscope_cyclenow();
```

Comment:
I tried to use functions in timers_preciseclock.h, but they are only valid if GRPC_TIMERS_RDTSC is defined (default is not). Endoscope must use a real time clock to preserve precision, regardless of what clock gRPC core uses.